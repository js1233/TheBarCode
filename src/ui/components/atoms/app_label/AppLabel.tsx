import { FONT_SIZE, FONT_SIZE_LINE_HEIGHT } from "config/Dimens";
import Fonts from "config/Fonts";
import { usePreferredTheme } from "hooks";
import React from "react";
import {
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TextStyle
} from "react-native";
import { TouchableOpacity } from "react-native-gesture-handler";

export interface AppLabelProps extends TextProps {
  text?: string;
  style?: StyleProp<TextStyle>;
  onPress?: () => void;
  textType?: TEXT_TYPE;
}

type Props = AppLabelProps;

export enum TEXT_TYPE {
  NORMAL = "normal",
  ITALIC = "italic",
  BOLD = "bold",
  UNDERLINE = "underline",
  LINE_THROUGH = "line-through",
  SEMI_BOLD = "semi-bold"
}

export const AppLabel = React.memo<Props>(
  ({ text, style, onPress, textType, ...rest }) => {
    const theme = usePreferredTheme();

    const getTextStyle = () => {
      if (textType === TEXT_TYPE.NORMAL) {
        return styles.normal;
      } else if (textType === TEXT_TYPE.ITALIC) {
        return styles.italic;
      } else if (textType === TEXT_TYPE.BOLD) {
        return styles.bold;
      } else if (textType === TEXT_TYPE.UNDERLINE) {
        return styles.underLine;
      } else if (textType === TEXT_TYPE.LINE_THROUGH) {
        return styles.lineThrough;
      } else if (textType === TEXT_TYPE.SEMI_BOLD) {
        return styles.semi_bold;
      } else {
        return styles.normal;
      }
    };

    const textJsx = (
      <Text
        style={[
          getTextStyle(),
          {
            color: theme.themedColors.interface["900"],
            fontSize: FONT_SIZE.sm
          },
          {
            lineHeight: FONT_SIZE_LINE_HEIGHT.ofFontSize(
              // @ts-ignore
              StyleSheet.flatten(style)?.fontSize ?? 16.0
            )
          },
          style
        ]}
        numberOfLines={1}
        {...rest}>
        {text}
      </Text>
    );

    if (onPress) {
      return (
        <TouchableOpacity activeOpacity={0.3} onPress={onPress}>
          {textJsx}
        </TouchableOpacity>
      );
    } else {
      return textJsx;
    }
  }
);

const styles = StyleSheet.create({
  bold: {
    fontFamily: Fonts.bold
  },
  normal: {
    fontFamily: Fonts.regular
  },
  semi_bold: {
    fontFamily: Fonts.semi_bold
  },
  italic: {
    fontStyle: "italic"
  },
  underLine: {
    textDecorationLine: "underline"
  },
  lineThrough: {
    textDecorationLine: "line-through"
  }
});
