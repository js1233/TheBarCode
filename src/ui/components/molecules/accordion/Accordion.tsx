import React, { useEffect, useState } from "react";
import { StyleSheet, TouchableOpacity, View } from "react-native";

type AccordionProps = {
  header: (isExpanded: boolean) => React.ReactElement;
  expandableItem: () => React.ReactElement | null;
  isExpanded?: boolean;
  executingTask?: () => boolean;
};

const Accordion: React.FC<AccordionProps> = ({
  header,
  expandableItem,
  isExpanded = false,
  executingTask
}) => {
  const [expanded, setExpanded] = useState(isExpanded);

  useEffect(() => {
    executingTask && setExpanded(executingTask());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <View style={style.container}>
      <TouchableOpacity
        testID="accordion-header"
        onPress={() => {
          setExpanded((prevState) => !prevState);
        }}>
        {header ? header(expanded) : null}
      </TouchableOpacity>
      <View testID="accordion-body">
        {expanded ? expandableItem() : null}
      </View>
    </View>
  );
};

export default Accordion;

const style = StyleSheet.create({
  container: {
    flex: 1
  }
});
