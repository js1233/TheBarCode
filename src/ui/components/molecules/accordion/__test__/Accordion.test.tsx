import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { TouchableOpacity } from "react-native";
import { AppLabel } from "ui/components/atoms/app_label/AppLabel";
import Accordion from "ui/components/molecules/accordion/Accordion";

test("snapshot testing", () => {
  const header = () => {
    return (
      <TouchableOpacity testID="header-click">
        <AppLabel text={"Sub Title"} />
      </TouchableOpacity>
    );
  };

  const body = () => {
    return <AppLabel text={"Sub Title"} />;
  };
  const rendered = render(
    <Accordion header={header} expandableItem={body} isExpanded={false} />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("renders header correctly", () => {
  const onPress = jest.fn();
  const header = () => {
    return (
      <TouchableOpacity onPress={onPress} testID="header-click">
        <AppLabel text={"Sub Title"} />
      </TouchableOpacity>
    );
  };

  const body = () => {
    return <AppLabel text={"Sub Title"} />;
  };
  const { queryAllByTestId } = render(
    <Accordion header={header} expandableItem={body} isExpanded={false} />
  );
  const headerView = queryAllByTestId("accordion-header");
  let bodyView = queryAllByTestId("accordion-body");
  expect(headerView).not.toBeNull();
  expect(bodyView).not.toBeNull();
});

it("accordion body should not null when click event occured", () => {
  const onPress = jest.fn();
  const header = () => {
    return (
      <TouchableOpacity onPress={onPress} testID="header-click">
        <AppLabel text={"Sub Title"} />
      </TouchableOpacity>
    );
  };

  const body = () => {
    return <AppLabel text={"Sub Title"} />;
  };
  const { queryByTestId } = render(
    <Accordion header={header} expandableItem={body} isExpanded={true} />
  );
  let bodyView = queryByTestId("accordion-body");

  const button = queryByTestId("header-click");
  if (button) {
    fireEvent.press(button);
  }
  expect(onPress).toBeCalledTimes(1);
  expect(bodyView).not.toBeNull();
});
