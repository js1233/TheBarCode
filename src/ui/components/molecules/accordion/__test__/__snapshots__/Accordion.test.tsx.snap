// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="accordion-header"
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="header-click"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontFamily": "ProximaNova-Regular",
            },
            Object {
              "color": "#171717",
              "fontSize": 27,
            },
            Object {
              "lineHeight": 42,
            },
            undefined,
          ]
        }
      >
        Sub Title
      </Text>
    </View>
  </View>
  <View
    testID="accordion-body"
  />
</View>
`;
